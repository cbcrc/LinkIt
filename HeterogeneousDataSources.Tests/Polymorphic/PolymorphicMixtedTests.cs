using System.Collections.Generic;
using ApprovalTests.Reporters;
using HeterogeneousDataSources.Tests.Shared;
using NUnit.Framework;
using RC.Testing;

namespace HeterogeneousDataSources.Tests.Polymorphic {
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class PolymorphicMixtedTests {
        private FakeReferenceLoader<Model, string> _fakeReferenceLoader;
        private LoadLinkProtocol _sut;

        [SetUp]
        public void SetUp() {
            var loadLinkProtocolBuilder = new LoadLinkProtocolBuilder();

            loadLinkProtocolBuilder.For<LinkedSource>()
                .IsRoot<string>()
                .PolymorphicLoadLink(
                    linkedSource => linkedSource.Model.TargetReference,
                    linkedSource => linkedSource.Target,
                    link => link.GetType(),
                    includes => includes
                        .WhenReference<Person,string>(
                            typeof(int),
                            link=>link.ToString()
                        )
                        .WhenNestedLinkedSource<PersonLinkedSource, string>(
                            typeof(string),
                            link => link.ToString()
                        )
                        .WhenSubLinkedSource<PersonLinkedSource, Person>(
                            typeof(Person)
                        )
                );
            loadLinkProtocolBuilder.For<PersonLinkedSource>()
                .LoadLinkReference(
                    linkedSource => linkedSource.Model.SummaryImageId,
                    linkedSource => linkedSource.SummaryImage
                );

            _fakeReferenceLoader =
                new FakeReferenceLoader<Model, string>(reference => reference.Id);
            _sut = loadLinkProtocolBuilder.Build(_fakeReferenceLoader);
        }

        [Test]
        public void LoadLink_MixedPolymorphicAsReference() {
            _fakeReferenceLoader.FixValue(
                new Model {
                    Id = "1",
                    TargetReference = 1
                }
            );

            var actual = _sut.LoadLink<LinkedSource,string>("1");

            ApprovalsExt.VerifyPublicProperties(actual);
        }

        [Test]
        public void LoadLink_MixedPolymorphicAsNestedLinkedSource() {
            _fakeReferenceLoader.FixValue(
                new Model {
                    Id = "1",
                    TargetReference = "nested"
                }
            );

            var actual = _sut.LoadLink<LinkedSource,string>("1");

            ApprovalsExt.VerifyPublicProperties(actual);
        }

        [Test]
        public void LoadLink_MixedPolymorphicAsSubLinkedSource() {
            _fakeReferenceLoader.FixValue(
                new Model {
                    Id = "1",
                    TargetReference = new Person{
                        Id = "as-sub-linked-source",
                        Name = "The Name",
                        SummaryImageId = "the-id"
                    }
                }
            );

            var actual = _sut.LoadLink<LinkedSource,string>("1");

            ApprovalsExt.VerifyPublicProperties(actual);
        }

        public class LinkedSource : ILinkedSource<Model> {
            public Model Model { get; set; }
            public object Target { get; set; }
        }

        public class Model{
            public string Id { get; set; }
            public object TargetReference { get; set; }
        }
    }
}
