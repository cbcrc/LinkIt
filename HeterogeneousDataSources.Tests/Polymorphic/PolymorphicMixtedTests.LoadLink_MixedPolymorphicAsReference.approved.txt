{
  "model": {
    "id": "1",
    "targetReference": 1
  },
  "target": {
    "$type": "HeterogeneousDataSources.Tests.Shared.Person, HeterogeneousDataSources.Tests",
    "id": "1",
    "name": "name-1",
    "summaryImageId": "person-img-1"
  }
}